File: js/3_1_1_strategy.js
class Context{
    constructor(strategy){
        this._strategy = strategy;
    }
        
    set strategy(value){
        if (!value || !value.doAlgorithm)
            throw "Bad strategy";
        this._strategy = value;
    }

    doSomeBusinessLogic(){
        console.log("Context: Sorting data using the strategy (not sure how it'll do it)");
        let data = [ "a", "b", "e", "c", "d", ];
        let result = this._strategy.doAlgorithm(data);

        let resultStr = result.join(",");

        console.log(resultStr);
    }
}

class ConcreteStrategyA {
    doAlgorithm(data){
        return data.sort();
    }
}

class ConcreteStrategyB{
    doAlgorithm(data){
        return data.sort().reverse();
    }
}
    
export { Context, ConcreteStrategyA, ConcreteStrategyB }; 

File: cs/3_1_1_Strategy.cs
using System;
using System.Collections.Generic;

namespace Behavioral.Strategy
{
    class Context
    {
        private IStrategy _strategy;

        public Context()
        { }

        public Context(IStrategy strategy)
        {
            this._strategy = strategy;
        }
        public void SetStrategy(IStrategy strategy)
        {
            this._strategy = strategy;
        }

        public void DoSomeBusinessLogic()
        {
            Console.WriteLine("Context: Sorting data using the strategy (not sure how it'll do it)");
            var data = new List<string> { "a", "b", "e", "c", "d", };
            var result = this._strategy.DoAlgorithm(data) as List<string>;

            string resultStr = String.Join(",", result);

            Console.WriteLine(resultStr);
        }
    }

    public interface IStrategy
    {
        object DoAlgorithm(object data);
    }

    class ConcreteStrategyA : IStrategy
    {
        public object DoAlgorithm(object data)
        {
            var list = data as List<string>;
            list.Sort();

            return list;
        }
    }

    class ConcreteStrategyB : IStrategy
    {
        public object DoAlgorithm(object data)
        {
            var list = data as List<string>;
            list.Sort();
            list.Reverse();

            return list;
        }
    }
}